File: packages/core/src/aggregations/periods.ts
/**
 * Period bucketing helpers.
 *
 * This is local-time math via the JS Date API: it respects the host's
 * timezone and honors the tracker's custom period window (`PeriodWindow`),
 * but doesn't yet support a tracker-specific timezone. When that lands, this
 * module is where it goes.
 */

import type { WeekStart } from '../schema/tables.js';

export type BucketPeriod = 'day' | 'week' | 'month' | 'year';

export interface Bucket {
  /** Inclusive lower bound, ISO 8601 local time. */
  start: string;
  /** Exclusive upper bound, ISO 8601 local time. */
  end: string;
  /** Stable identifier, e.g. "2026-W21" for weeks, "2026-05" for months. */
  label: string;
}

/**
 * Where a tracker's periods begin. The field names match the tracker columns
 * they come from, so a `Tracker` can be passed straight in; every field is
 * optional and falls back to plain calendar bucketing.
 */
export interface PeriodWindow {
  /**
   * Minutes after local midnight a day begins: 420 runs a day 7:00 AM → 6:59
   * AM. Shifts every period, since all of them start on a day.
   */
  day_start_minute?: number;
  /** Weekday a week begins on: 0 = Sunday, 1 = Monday. */
  week_start?: WeekStart;
  /**
   * Day-of-month a month begins on, 1..28: 8 runs a month the 8th → the 7th
   * of the next month. Also the day a year begins on.
   */
  month_start_day?: number;
  /** Month a year begins on, 1..12: 4 runs a year April → March. */
  year_start_month?: number;
}

interface ResolvedWindow {
  dayStartMinute: number;
  weekStart: WeekStart;
  monthStartDay: number;
  yearStartMonth: number;
}

/** Fill in the calendar defaults and drop out-of-range components. */
function resolve(window: PeriodWindow): ResolvedWindow {
  const inRange = (v: number | undefined, min: number, max: number, fallback: number): number =>
    v !== undefined && Number.isInteger(v) && v >= min && v <= max ? v : fallback;
  return {
    dayStartMinute: inRange(window.day_start_minute, 0, 1439, 0),
    weekStart: (inRange(window.week_start, 0, 1, 1) as WeekStart),
    monthStartDay: inRange(window.month_start_day, 1, 28, 1),
    yearStartMonth: inRange(window.year_start_month, 1, 12, 1),
  };
}

/**
 * Start of the *logical* day containing `instant` — local midnight plus the
 * window's day start, stepping back a calendar day for an instant that falls
 * before the day has opened (03:30 with a 7:00 AM day start still belongs to
 * the previous day).
 */
function dayStartOf(instant: Date, w: ResolvedWindow): Date {
  const d = new Date(instant);
  d.setHours(0, w.dayStartMinute, 0, 0);
  if (instant.getHours() * 60 + instant.getMinutes() < w.dayStartMinute) {
    d.setDate(d.getDate() - 1);
  }
  return d;
}

/** Start of the bucket containing `instant`. */
export function bucketStart(
  instant: Date,
  period: BucketPeriod,
  window: PeriodWindow = {},
): Date {
  const w = resolve(window);
  // Every period opens at a day boundary, so start from the logical day and
  // walk back to the period's first day.
  const d = dayStartOf(instant, w);

  switch (period) {
    case 'day':
      return d;
    case 'week': {
      const dow = d.getDay(); // 0 = Sunday ... 6 = Saturday
      const diff = (dow - w.weekStart + 7) % 7;
      d.setDate(d.getDate() - diff);
      return d;
    }
    case 'month': {
      // Before this month's opening day, we're still in the window that opened
      // last month; setMonth normalizes month -1 to last December.
      if (d.getDate() < w.monthStartDay) d.setMonth(d.getMonth() - 1, 1);
      d.setDate(w.monthStartDay);
      return d;
    }
    case 'year': {
      const start = new Date(d);
      start.setMonth(w.yearStartMonth - 1, w.monthStartDay);
      if (d < start) start.setFullYear(start.getFullYear() - 1);
      return start;
    }
  }
}

/** Start of the next bucket after the one containing `instant`. */
export function bucketEnd(
  instant: Date,
  period: BucketPeriod,
  window: PeriodWindow = {},
): Date {
  const start = bucketStart(instant, period, window);
  const end = new Date(start);
  switch (period) {
    case 'day':
      end.setDate(end.getDate() + 1);
      return end;
    case 'week':
      end.setDate(end.getDate() + 7);
      return end;
    case 'month':
      // The date is month_start_day (<= 28), so the next month always has it.
      end.setMonth(end.getMonth() + 1);
      return end;
    case 'year':
      end.setFullYear(end.getFullYear() + 1);
      return end;
  }
}

/**
 * Stable identifier for a bucket starting at `start`. With a custom window the
 * label names the calendar date/month/year the bucket *opens* in: a month
 * window running May 8 → June 7 is labelled "2026-05".
 */
export function bucketLabel(start: Date, period: BucketPeriod): string {
  const yyyy = start.getFullYear();
  const mm = String(start.getMonth() + 1).padStart(2, '0');
  const dd = String(start.getDate()).padStart(2, '0');
  switch (period) {
    case 'day':
      return `${yyyy}-${mm}-${dd}`;
    case 'week':
      return `${yyyy}-W${String(isoWeekNumber(start)).padStart(2, '0')}`;
    case 'month':
      return `${yyyy}-${mm}`;
    case 'year':
      return String(yyyy);
  }
}

/** ISO 8601 week number (1..53). Used only for labelling. */
function isoWeekNumber(date: Date): number {
  // Copy and shift to Thursday of the same week (ISO weeks are anchored on Thursday).
  const d = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()));
  const day = d.getUTCDay() || 7;
  d.setUTCDate(d.getUTCDate() + 4 - day);
  const yearStart = new Date(Date.UTC(d.getUTCFullYear(), 0, 1));
  const weekNo = Math.ceil(((d.getTime() - yearStart.getTime()) / 86_400_000 + 1) / 7);
  return weekNo;
}

File: server/internal/core/periods.go
package core

import (
	"fmt"
	"time"
)

// Period bucketing helpers, ported from aggregations/periods.ts.
//
// This is host-local-time math: it respects the server's timezone, honors the
// tracker's custom period window (PeriodWindow), and doesn't yet support a
// tracker-specific timezone. When that lands, this file is where it goes.

// BucketPeriod is one of day, week, month, year.
type BucketPeriod string

const (
	PeriodDay   BucketPeriod = "day"
	PeriodWeek  BucketPeriod = "week"
	PeriodMonth BucketPeriod = "month"
	PeriodYear  BucketPeriod = "year"
)

// ValidBucketPeriod reports whether s names a bucket period.
func ValidBucketPeriod(s string) bool {
	switch BucketPeriod(s) {
	case PeriodDay, PeriodWeek, PeriodMonth, PeriodYear:
		return true
	}
	return false
}

// PeriodWindow is a tracker's definition of where its periods begin: the four
// tracker columns that shift bucket boundaries off the plain calendar.
//
// The zero value is not meaningful — build one with NewPeriodWindow (or
// windowOf for a tracker), which fills in the schema defaults and clamps
// out-of-range values, so a window built from a hand-written row still
// produces calendar buckets rather than nonsense.
type PeriodWindow struct {
	// Minutes after local midnight a day begins: 420 runs a day 7:00 AM →
	// 6:59 AM. Shifts every period, since all of them start on a day.
	DayStartMinute int
	// Weekday a week begins on: 0 = Sunday, 1 = Monday.
	WeekStart int
	// Day-of-month a month begins on, 1..28: 8 runs a month the 8th → the
	// 7th of the next month. Also the day a year begins on.
	MonthStartDay int
	// Month a year begins on, 1..12: 4 runs a year April → March.
	YearStartMonth int
}

// DefaultPeriodWindow is plain calendar bucketing: midnight, Monday weeks, the
// 1st of the month, January. It matches the columns' SQL defaults.
var DefaultPeriodWindow = PeriodWindow{DayStartMinute: 0, WeekStart: 1, MonthStartDay: 1, YearStartMonth: 1}

// NewPeriodWindow builds a window, replacing out-of-range components with the
// calendar default.
func NewPeriodWindow(dayStartMinute, weekStart, monthStartDay, yearStartMonth int) PeriodWindow {
	w := DefaultPeriodWindow
	if dayStartMinute >= 0 && dayStartMinute <= 1439 {
		w.DayStartMinute = dayStartMinute
	}
	if weekStart == 0 || weekStart == 1 {
		w.WeekStart = weekStart
	}
	if monthStartDay >= 1 && monthStartDay <= 28 {
		w.MonthStartDay = monthStartDay
	}
	if yearStartMonth >= 1 && yearStartMonth <= 12 {
		w.YearStartMonth = yearStartMonth
	}
	return w
}

// windowOf reads a tracker's period window; a missing tracker buckets by the
// plain calendar.
func windowOf(t *Tracker) PeriodWindow {
	if t == nil {
		return DefaultPeriodWindow
	}
	return NewPeriodWindow(t.DayStartMinute, t.WeekStart, t.MonthStartDay, t.YearStartMonth)
}

// dayStart returns the start of the *logical* day containing instant — local
// midnight plus DayStartMinute, stepping back a calendar day for an instant
// that falls before the day has opened (03:30 with a 7:00 AM day start still
// belongs to the previous day).
func (w PeriodWindow) dayStart(instant time.Time) time.Time {
	l := instant.In(time.Local)
	y, m, d := l.Date()
	if l.Hour()*60+l.Minute() < w.DayStartMinute {
		d--
	}
	return time.Date(y, m, d, 0, w.DayStartMinute, 0, 0, time.Local)
}

// bucketStart returns the local-time start of the bucket containing instant.
func bucketStart(instant time.Time, period BucketPeriod, w PeriodWindow) time.Time {
	// Every period opens at a day boundary, so start from the logical day and
	// walk back to the period's first day. Date() below therefore reads the
	// logical day's calendar date, not the raw instant's.
	day := w.dayStart(instant)
	y, m, d := day.Date()

	switch period {
	case PeriodDay:
		return day
	case PeriodWeek:
		dow := int(day.Weekday()) // 0 = Sunday … 6 = Saturday
		diff := (dow - w.WeekStart + 7) % 7
		return w.at(y, m, d-diff)
	case PeriodMonth:
		// Before this month's opening day, we're still in the window that
		// opened last month. time.Date normalizes month 0 to last December.
		if d < w.MonthStartDay {
			m--
		}
		return w.at(y, m, w.MonthStartDay)
	case PeriodYear:
		start := w.at(y, time.Month(w.YearStartMonth), w.MonthStartDay)
		if day.Before(start) {
			start = w.at(y-1, time.Month(w.YearStartMonth), w.MonthStartDay)
		}
		return start
	}
	return day
}

// at builds a bucket boundary: the given local date at the window's day-start
// minute. Out-of-range components normalize as time.Date does (month 0 = last
// December, day 0 = last day of the previous month).
func (w PeriodWindow) at(y int, m time.Month, d int) time.Time {
	return time.Date(y, m, d, 0, w.DayStartMinute, 0, 0, time.Local)
}

// bucketEnd returns the start of the next bucket after the one containing
// instant.
func bucketEnd(instant time.Time, period BucketPeriod, w PeriodWindow) time.Time {
	start := bucketStart(instant, period, w)
	y, m, d := start.Date()
	switch period {
	case PeriodDay:
		return w.at(y, m, d+1)
	case PeriodWeek:
		return w.at(y, m, d+7)
	case PeriodMonth:
		// d is MonthStartDay (≤ 28), so the next month always has it.
		return w.at(y, m+1, d)
	case PeriodYear:
		return w.at(y+1, m, d)
	}
	return start
}

// bucketLabel renders the stable identifier for a bucket starting at start,
// e.g. "2026-05-25", "2026-W21", "2026-05", "2026". With a custom window the
// label names the calendar date/month/year the bucket *opens* in: a month
// window running May 8 → June 7 is labelled "2026-05".
func bucketLabel(start time.Time, period BucketPeriod) string {
	l := start.In(time.Local)
	switch period {
	case PeriodDay:
		return l.Format("2006-01-02")
	case PeriodWeek:
		return l.Format("2006") + "-W" + pad2(isoWeekNumber(l))
	case PeriodMonth:
		return l.Format("2006-01")
	case PeriodYear:
		return l.Format("2006")
	}
	return l.Format("2006-01-02")
}

func pad2(n int) string { return fmt.Sprintf("%02d", n) }

// isoWeekNumber ports the TS helper: ISO 8601 week number (1..53) computed
// from the local calendar date. Note the original derives it from the *local*
// year — labels match the TS output exactly, including its week-53/January
// quirk near year boundaries.
func isoWeekNumber(date time.Time) int {
	y, m, d := date.Date()
	// Shift to Thursday of the same week (ISO weeks are anchored on Thursday).
	dt := time.Date(y, m, d, 0, 0, 0, 0, time.UTC)
	day := int(dt.Weekday())
	if day == 0 {
		day = 7
	}
	dt = dt.AddDate(0, 0, 4-day)
	yearStart := time.Date(dt.Year(), 1, 1, 0, 0, 0, 0, time.UTC)
	weekNo := (int(dt.Sub(yearStart).Hours()/24) + 1 + 6) / 7
	return weekNo
}
